dvc/checkout.py

```diff
@@ -169,135 +130,109 @@ def _checkout_file(
-            modified = True
-            _remove(path_info, fs, cache, force=force)
-            _link(cache, cache_info, path_info)
-        else:
+    cache_info = cache.hash_to_path_info(change.new.obj.hash_info.value)
+    if change.old.obj:
+        if relink:
             if fs.iscopy(path_info) and _cache_is_copy(cache, path_info):
                 cache.unprotect(path_info)
             else:
-                _remove(path_info, fs, cache, force=force)
+                _remove(path_info, fs, change.old.in_cache, force=force)
                 _link(cache, cache_info, path_info)
+        else:
+            modified = True
+            _remove(path_info, fs, change.old.in_cache, force=force)
+            _link(cache, cache_info, path_info)
     else:
         _link(cache, cache_info, path_info)
         modified = True
 
     if state:
-        state.save(path_info, fs, obj.hash_info)
+        state.save(path_info, fs, change.new.obj.hash_info)
 
     if progress_callback:
         progress_callback(str(path_info))
 
     return modified
 
 
-def _remove_redundant_files(
-    path_info, fs, obj, cache, force, dvcignore: Optional[DvcIgnoreFilter]
+def _diff(
+    path_info,
+    fs,
+    obj,
+    cache,
+    relink=False,
+    dvcignore: Optional[DvcIgnoreFilter] = None,
 ):
-    if dvcignore:
-        existing_files = set(dvcignore.walk_files(fs, path_info))
-    else:
-        existing_files = set(fs.walk_files(path_info))
+    old = None
+    try:
+        _, old = stage(
+            cache,
+            path_info,
+            fs,
+            obj.hash_info.name if obj else cache.fs.PARAM_CHECKSUM,
+            dry_run=True,
+            dvcignore=dvcignore,
+        )
+    except FileNotFoundError:
+        pass
 
-    needed_files = {path_info.joinpath(*key) for key, _ in obj}
-    redundant_files = existing_files - needed_files
-    for path in redundant_files:
-        _remove(path, fs, cache, force)
+    diff = odiff(old, obj, cache)
 
-    return bool(redundant_files)
+    if relink:
+        diff.modified.extend(diff.unchanged)
 
+    return diff
 
-def _checkout_dir(
+
+def _checkout(
+    diff,
     path_info,
     fs,
-    obj,
     cache,
-    force,
+    force=False,
     progress_callback=None,
     relink=False,
-    dvcignore: Optional[DvcIgnoreFilter] = None,
     state=None,
 ):
-    modified = False
-    # Create dir separately so that dir is created
-    # even if there are no files in it
-    if not fs.exists(path_info):
-        modified = True
-        fs.makedirs(path_info)
+    if not diff:
+        return
 
-    logger.debug("Linking directory '%s'.", path_info)
+    for change in diff.deleted:
+        entry_path = (
+            path_info.joinpath(*change.old.key)
+            if change.old.key != ROOT
+            else path_info
+        )
+        _remove(entry_path, fs, change.old.in_cache, force=force)
 
     failed = []
-    for entry_key, entry_obj in obj:
+    for change in diff.added + diff.modified:
```

Comment:
concatenation might be expensive.
```suggestion
    for change in chain(diff.added, diff.modified):
```

Reply:
Nice catch!